### PR TITLE
ESQL: Fix sorts containing _source

### LINKED ISSUE #116659: Suppressed rest error for "No TopN sorting encoder for type SOURCE"
Related to a previously fixed issue at https://github.com/elastic/elasticsearch/issues/114423, we are seeing errors for `No TopN sorting encoder for type SOURCE` which should not occur since we should never SORT on _source. We do not know the query causing this, but the stack trace is available, which might help track this down.

```
org.elasticsearch.xpack.esql.EsqlIllegalArgumentException: No TopN sorting encoder for type SOURCE
	at org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.planTopN(LocalExecutionPlanner.java:360)
	at org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:195)
	at org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.planOutput(LocalExecutionPlanner.java:276)
	at org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:233)
	at org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:174)
	at org.elasticsearch.xpack.esql.plugin.ComputeService.runCompute(ComputeService.java:446)
	at org.elasticsearch.xpack.esql.plugin.ComputeService.execute(ComputeService.java:214)
	at org.elasticsearch.xpack.esql.plugin.TransportEsqlQueryAction.lambda$innerExecute$3(TransportEsqlQueryAction.java:176)
	at org.elasticsearch.xpack.esql.session.EsqlSession.executeSubPlans(EsqlSession.java:211)
	at org.elasticsearch.xpack.esql.session.EsqlSession.executeOptimizedPlan(EsqlSession.java:176)
	at org.elasticsearch.xpack.esql.session.EsqlSession$1.onResponse(EsqlSession.java:155)
	at org.elasticsearch.xpack.esql.session.EsqlSession$1.onResponse(EsqlSession.java:152)
	at org.elasticsearch.xpack.esql.session.EsqlSession.lambda$preAnalyze$13(EsqlSession.java:338)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:247)
	at org.elasticsearch.xpack.esql.session.IndexResolver.lambda$resolveAsMergedMapping$0(IndexResolver.java:82)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:247)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:203)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:197)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onResponse(ActionListenerImplementations.java:336)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:33)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:97)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.fieldcaps.TransportFieldCapabilitiesAction.mergeIndexResponses(TransportFieldCapabilitiesAction.java:325)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.fieldcaps.TransportFieldCapabilitiesAction.lambda$doExecuteForked$6(TransportFieldCapabilitiesAction.java:230)
	at org.elasticsearch.base@9.0.0/org.elasticsearch.core.AbstractRefCounted$1.closeInternal(AbstractRefCounted.java:125)
	at org.elasticsearch.base@9.0.0/org.elasticsearch.core.AbstractRefCounted.decRef(AbstractRefCounted.java:77)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.support.RefCountingRunnable.close(RefCountingRunnable.java:113)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.RunOnce.run(RunOnce.java:41)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.fieldcaps.RequestDispatcher.innerExecute(RequestDispatcher.java:159)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.fieldcaps.RequestDispatcher$1.doRun(RequestDispatcher.java:128)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:34)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1023)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1575)
```

### PR DESCRIPTION
**Pull request by nik9000**
This fixes sorts containing the a `_source` field. It can use the standard encoder for `BytesRef`s. You can't sort *by* a `_source` field, but that doesn't really make sense ayway.

Closes #116659

### CODE REVIEW

**Review by alex-spies — APPROVED** [review]
LGTM except for a test remark which could make bwc tests break if I'm correct.

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/140_metadata.yml [hunk 1]
```diff
@@ -170,4 +170,43 @@ setup:
...
+
+---
+"sort returning _source is allowed":
+  - requires:
+      test_runner_features: [capabilities]
+      capabilities:
+        - method: POST
+          path: /_query
+          parameters: []
+          capabilities: [sorting_on_source_and_counters_forbidden]
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/140_metadata.yml:183]: Oh yeah, we will. And that's why it's failing....

[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/140_metadata.yml:183]: > I'm confused why CI didn't fail for this reason (did we just get lucky in this run?).

If failed for 8.17. And we don't test 8.16 against main at the moment.

[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/140_metadata.yml:183]: Or maybe the 8.17 failed for another eason....

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/140_metadata.yml [hunk 2]
```diff
@@ -170,4 +170,43 @@ setup:
...
+      reason: "Sorting on _source shouldn't have been possible"
+  - do:
+      esql.query:
+        body:
+          query: 'FROM test metadata _source | SORT case ASC | KEEP case, _source | LIMIT 5'
```
Comment (bpintea) [inline x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/140_metadata.yml:188]: nitty nit, for a next PR that might touch this (here and above)
```suggestion
          query: 'FROM test METADATA _source | SORT case ASC | KEEP case, _source | LIMIT 5'
```

**Comment by nik9000** [comment]
@elasticmachine, update branch

**Pull request merged**